File: app/src/main/java/com/example/liftoff/extractor/email/EmailExtractor.java
package com.example.liftoff.extractor.email;

import com.example.liftoff.extractor.Extractor;
import com.example.liftoff.storage.filesystem.FileStorage;

/**
 * A generic abstract extractor class, that holds
 * the typical email extraction methods.
 */
public abstract class EmailExtractor implements Extractor {
    protected boolean isAuthenticated = false;
    protected final String emailAddress;
    protected final String userToken;
    protected final FileStorage storage;

    protected EmailExtractor(final String emailAddress, final String userToken, FileStorage storage) {
        this.emailAddress = emailAddress;
        this.userToken = userToken;
        this.storage = storage;
    }

    public abstract void authenticate() throws Exception;

    public abstract void close() throws Exception;
}

File: src/main/scala/com/example/liftoff/extractor/email/EmailExtractor.scala
package com.example.liftoff.extractor.email

import com.example.liftoff.error.ExtractorError
import com.example.liftoff.extractor.Extractor
import com.example.liftoff.storage.Storage
import zio.IO

abstract class EmailExtractor(emailAddress: String, userToken: String, storage: Storage) extends Extractor {
  protected var isAuthenticated: Boolean = false

  def authenticate(): IO[ExtractorError, Unit]

  def close(): IO[ExtractorError, Unit]
}
